dali/imgcodec/future_decode_results.cc

```diff
@@ -0,0 +1,206 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <deque>
+#include <mutex>
+#include <condition_variable>
+#include "dali/imgcodec/image_decoder_interfaces.h"
+
+namespace dali {
+namespace imgcodec {
+
+class DecodeResultsSharedState {
+ public:
+  static std::shared_ptr<DecodeResultsSharedState> get() {
+    if (free_.empty())
+      return std::make_shared<DecodeResultsSharedState>();
```

Comment:
I'd add an explanation somewhere about the motivation behind `free_` (we want to avoid expensive resource allocation, etc)